src/casync-http.c

```diff
@@ -11,67 +12,1066 @@
+                }
+
+                /* At this point, handle and chunk data are left "unconfigured"
+                 * in the ready queue. They'll be reconfigured when re-used. */
+                queue_push(dl->ready, handle);
+        }
+
+        return i;
+}
+
+/* Process chunks that were downloaded by curl, return the number of chunks handled */
+static int ca_chunk_downloader_process_curl_multi(CaChunkDownloader *dl) {
+        int i, n;
+        CURLMcode cm;
+
+        cm = curl_multi_perform(dl->multi, &n);
+        if (cm != CURLM_OK) {
+                log_error_curlm(cm, "Failed to perform curl multi");
+                return -EIO;
+        }
+
+        for (i = 0; ; i++) {
+                CURLcode c;
+                CURLMsg *msg;
+                CURL *handle;
+                long protocol_status;
+                const char *effective_url;
+                ChunkData *cd;
+
+                msg = curl_multi_info_read(dl->multi, &n);
+                if (!msg)
+                        break;
+
+                if (msg->msg != CURLMSG_DONE) {
+                        log_error("Unexpected CURL message: %d", msg->msg);
+                        return -EIO;
+                }
+
+                if (msg->data.result != CURLE_OK) {
+                        log_error_curle(msg->data.result, "Failed to acquire chunk");
+                        return -EIO;
+                }
+
+                handle = msg->easy_handle;
+
+                effective_url = get_curl_effective_url(handle);
+                if (!effective_url)
+                        return -EIO;
+
+                cd = get_curl_private(handle);
+                if (!cd)
+                        return -EIO;
+
+                c = curl_easy_getinfo(handle, CURLINFO_RESPONSE_CODE, &protocol_status);
+                if (c != CURLE_OK) {
+                        log_error_curle(c, "Failed to query response code");
+                        return -EIO;
+                }
+
+                if (!protocol_status_ok(arg_protocol, protocol_status)) {
+                        log_error("%s server failure %ld while requesting %s",
+                                  protocol_str(arg_protocol), protocol_status,
+                                  effective_url);
+
+                        // TODO support multiple stores
+
+                        /* No more stores? Set current_store to a special value
+                         * to indicate failure. */
+                        cd->current_store = SIZE_MAX;
+                }
+
+                cm = curl_multi_remove_handle(dl->multi, handle);
+                if (cm != CURLM_OK) {
+                        log_error_curlm(cm, "Failed to remove curl handle");
+                        return -EIO;
+                }
+
+                queue_remove(dl->inprogress, handle);
+                queue_push(dl->completed, handle);
+        }
+
+        return i;
+}
+
+static int ca_chunk_downloader_step(CaChunkDownloader *dl) {
+        int r;
```

Comment:
So, this function is tricky. This is basically what happens during a loop iteration.

See the call `ca_chunk_downloader_remote_step()` below? If you move it to the end, then you break the testsuite, and a bunch of `casync list` commands hang forever. I don't really know why. But it's not sure you will reproduce it. It's interesting to see that if I run the testuite without parallel tests (ie. `MESON_TESTTHREADS=5 ninja test`), then everything works fine.

So there's some subtleties going here that I didn't fully understand, and I ended up ordering things in this function by trial and failure. In other words, this needs thorough review: this function, the functions it calls, and basically the communication with the casync remote.